src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNNQuantizationStateReader.java

```diff
@@ -0,0 +1,155 @@
+        SegmentReadState segmentReadState = readConfig.getSegmentReadState();
+        String field = readConfig.getField();
+        String quantizationStateFileName = getQuantizationStateFileName(segmentReadState);
+        int fieldNumber = segmentReadState.fieldInfos.fieldInfo(field).getFieldNumber();
+
+        try (IndexInput input = segmentReadState.directory.openInput(quantizationStateFileName, IOContext.READ)) {
+            CodecUtil.retrieveChecksum(input);
+            int numFields = getNumFields(input);
+
+            long position = -1;
+            int length = 0;
+
+            // Read each field's metadata from the index section, break when correct field is found
+            for (int i = 0; i < numFields; i++) {
+                int tempFieldNumber = input.readInt();
+                int tempLength = input.readInt();
+                long tempPosition = input.readVLong();
+                if (tempFieldNumber == fieldNumber) {
+                    position = tempPosition;
+                    length = tempLength;
+                    break;
+                }
+            }
+
+            if (position == -1 || length == 0) {
+                throw new IllegalArgumentException(String.format("Field %s not found", field));
+            }
+
+            byte[] stateBytes = readStateBytes(input, position, length);
+
+            // Deserialize the byte array to a quantization state object
+            ScalarQuantizationType scalarQuantizationType = ((ScalarQuantizationParams) readConfig.getQuantizationParams()).getSqType();
+            if (scalarQuantizationType == ScalarQuantizationType.ONE_BIT) {
+                return OneBitScalarQuantizationState.fromByteArray(stateBytes);
+            } else if (scalarQuantizationType == ScalarQuantizationType.TWO_BIT
+                || scalarQuantizationType == ScalarQuantizationType.FOUR_BIT) {
+                    return MultiBitScalarQuantizationState.fromByteArray(stateBytes);
+                } else {
+                    throw new IllegalArgumentException(String.format("Unexpected scalar quantization type: %s", scalarQuantizationType));
+                }
+        }
+    }
+
+    @VisibleForTesting
+    static int getNumFields(IndexInput input) throws IOException {
+        long footerStart = input.length() - CodecUtil.footerLength();
+        long markerAndIndexPosition = footerStart - Integer.BYTES - Long.BYTES;
+        input.seek(markerAndIndexPosition);
+        long indexStartPosition = input.readLong();
+        input.readInt();
+        input.seek(indexStartPosition);
+        return input.readInt();
+    }
+
+    @VisibleForTesting
+    static byte[] readStateBytes(IndexInput input, long position, int length) throws IOException {
+        input.seek(position);
+        byte[] stateBytes = new byte[length];
+        input.readBytes(stateBytes, 0, length);
+        return stateBytes;
+    }
+
+    @VisibleForTesting
```

Comment:
why we are enabling @VisibleForTesting on these functions? These should be tested along with read functions

Reply:
Visible for testing doesn't actually do anything as it is just an annotation.  These methods would otherwise be private, but since they need to be tested I made them package level access.